packages/block-library/src/gallery/transforms.js

```diff
@@ -33,6 +34,98 @@ const parseShortcodeIds = ( ids ) => {
 	return ids.split( ',' ).map( ( id ) => parseInt( id, 10 ) );
 };
 
+/**
+ * Third party block plugins don't have an easy way to detect if the
+ * innerBlocks version of the Gallery is running when they run a
+ * 3rdPartyBlock -> GallaryBlock transform so this tranform filter
+ * will handle this. Once the innerBlocks version is the default
+ * in a core release, this could be deprecated and removed after
+ * plugin authors have been given time to update transforms.
+ *
+ * @typedef  {Object} Attributes
+ * @typedef  {Object} Block
+ * @property {Attributes} attributes The attributes of the block.
+ * @param    {Block}      block      The transformed block.
+ * @return   {Block}                 The transformed block.
+ */
+function updateThirdPartyTransformToGallery( block ) {
+	const settings = select( blockEditorStore ).getSettings();
+	if (
+		settings.__unstableGalleryWithImageBlocks &&
+		block.name === 'core/gallery' &&
+		block.attributes?.images
+	) {
+		const innerBlocks = block.attributes.images.map(
+			( { url, id, alt } ) => {
+				return createBlock( 'core/image', {
+					url,
+					id: parseInt( id, 10 ),
+					alt,
+					sizeSlug: block.attributes.sizeSlug,
+					linkDestination: block.attributes.linkDestination,
+				} );
+			}
+		);
+
+		delete block.attributes.ids;
+		delete block.attributes.images;
+		block.innerBlocks = innerBlocks;
+	}
+
+	return block;
+}
+addFilter(
+	'blocks.switchToBlockType.transformedBlock',
+	'core/gallery/update-third-party-transform-to',
+	updateThirdPartyTransformToGallery
+);
+
+/**
+ * Third party block plugins don't have an easy way to detect if the
+ * innerBlocks version of the Gallery is running when they run a
+ * 3rdPartyBlock -> GallaryBlock transform so this tranform filter
+ * will handle this. Once the innerBlocks version is the default
+ * in a core release, this could be deprecated and removed after
+ * plugin authors have been given time to update transforms.
+ *
+ * @typedef  {Object} Attributes
+ * @typedef  {Object} Block
+ * @property {Attributes} attributes The attributes of the block.
+ * @param    {Block}      toBlock    The block yo transform to.
+ * @param    {Block[]}    fromBlocks The block yo transform to.
+ * @return   {Block}                 The transformed block.
+ */
+function updateThirdPartyTransformFromGallery( toBlock, fromBlocks ) {
+	const settings = select( blockEditorStore ).getSettings();
+	const galleryBlock = fromBlocks.find(
+		( transformedBlock ) =>
+			transformedBlock.name === 'core/gallery' &&
+			transformedBlock.innerBlocks.length > 0 &&
+			! transformedBlock.attributes.images?.length > 0 &&
+			! toBlock.name.includes( 'core/' )
+	);
+
+	if ( settings.__unstableGalleryWithImageBlocks && galleryBlock ) {
+		const images = galleryBlock.innerBlocks.map(
+			( { attributes: { url, id, alt } } ) => ( {
+				url,
+				id: parseInt( id, 10 ),
```

Comment:
Might be a silly question, but is `id` ever `undefined`? If so, will this result in `id` being set to `NaN`. I _think_ we always have an id, but just thought I'd double check!

Reply:
Now you mention it, there are some times when the the id is null, eg if you add an image block and add the image from url option and not from upload or media browser and then transform the image block to a gallery. In case of core image block it doesn't get to this filter as handled properly by own transform, by added a check in just in case any third party blocks have same approach.

Reply:
Sounds good. Thanks Glen!